### PR TITLE
In GrainId, use concrete types instead of string and store the MembershipVersion

### PR DESCRIPTION
**Pull request by benjaminpetit**
The field `MembershipVersion` in the `GrainId` struct will help with address invalidation

### CODE REVIEW

File: src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs [hunk 1]
```diff
@@ -89,11 +89,24 @@ public async Task<GrainAddress> Register(GrainAddress address)
...
-            var value = JsonConvert.SerializeObject(address);
+            var value = JsonSerializer.Serialize(address);
 
             try
             {
-                var result = (int) await this.database.ScriptEvaluateAsync(this.deleteScript, new { key = GetKey(address.GrainId), val = JsonConvert.SerializeObject(address) });
+                if (address.MembershipVersion == MembershipVersion.MinValue)
+                {
+                    var fullAddress = await Lookup(address.GrainId);
+                    if (!fullAddress.Equals(address))
```
Comment (ReubenBond) [inline src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs:99]: It's unclear that `GrainAddress.Equals` does not compare the `MembershipVersion` property

File: src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs [hunk 2]
```diff
@@ -89,11 +89,24 @@ public async Task<GrainAddress> Register(GrainAddress address)
 
         public async Task Unregister(GrainAddress address)
         {
-            var value = JsonConvert.SerializeObject(address);
+            var value = JsonSerializer.Serialize(address);
 
             try
             {
-                var result = (int) await this.database.ScriptEvaluateAsync(this.deleteScript, new { key = GetKey(address.GrainId), val = JsonConvert.SerializeObject(address) });
+                if (address.MembershipVersion == MembershipVersion.MinValue)
```
Comment (ReubenBond) [inline src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs:96]: There should be a comment here explaining that this indicates that the provided `address` was not returned from the directory and must have been constructed externally - or some kind of comment.

Generally, we should aim for more descriptive comments.

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs [hunk 3]
```diff
@@ -54,9 +52,9 @@ public async ValueTask<ActivationAddress> Lookup(GrainId grainId)
                 return cachedResult;
             }
 
-            var entry = await GetGrainDirectory(grainId.Type).Lookup(grainId.ToString());
+            var entry = await GetGrainDirectory(grainId.Type).Lookup(grainId);
 
-            // Nothing found
+            // Nothing foundD
```
Comment (ReubenBond) [inline src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:57]:
```suggestion
            // Nothing found
```

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs [hunk 4]
```diff
@@ -105,7 +104,7 @@ public async Task<ActivationAddress> Register(ActivationAddress address)
             }
 
             // Cache update
-            this.cache.AddOrUpdate(activationAddress, 0);
+            this.cache.AddOrUpdate(activationAddress, (int) result.MembershipVersion.Value);
```
Comment (ReubenBond) [inline src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:107]: Should we change `cache` to hold `MembershipVersion` instances?

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs [hunk 5]
```diff
@@ -169,26 +168,17 @@ async Task OnStop(CancellationToken ct)
         private async Task ListenToClusterChange()
         {
             var previousSnapshot = this.clusterMembershipService.CurrentSnapshot;
-            // Update the list of known dead silos for lazy filtering for the first time
-            this.knownDeadSilos = new HashSet<SiloAddress>(previousSnapshot.Members.Values
-                .Where(m => m.Status == SiloStatus.Dead)
-                .Select(m => m.SiloAddress));
```
Comment (ReubenBond) [inline src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:175]: Very glad that we can delete all of this code

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs [hunk 6]
```diff
@@ -205,6 +195,25 @@ await foreach (var snapshot in updates)
             }
         }
 
+        private bool IsPointingToDeadSilo(GrainAddress grainAddress)
+            => IsPointingToDeadSilo(grainAddress.SiloAddress, grainAddress.MembershipVersion);
+
+        private bool IsPointingToDeadSilo(SiloAddress siloAddress, MembershipVersion membershipVersion)
```
Comment (ReubenBond) [inline src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:201]: Maybe rename to `IsKnownDeadSilo` to be clear that if this returns `true` then the silo is known to be dead but if it returns false then we do not know whether the silo is dead or not

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs [hunk 7]
```diff
@@ -205,6 +195,25 @@ await foreach (var snapshot in updates)
...
+            var current = this.clusterMembershipService.CurrentSnapshot;
+
+            // Check if the target silo is in the cluster
+            if (current.Members.TryGetValue(siloAddress, out var value))
+            {
+                // It is, check if it's alive
+                return value.Status.IsTerminating();
+            }
+
+            // We didn't found it in the cluster. If the silo entry is too old, it has been cleaned in the membership table: the entry isn't valid anymore.
```
Comment (ReubenBond) [inline src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:212]:
```suggestion
            // We didn't find it in the cluster. If the silo entry is too old, it has been cleaned in the membership table: the entry isn't valid anymore.
```

File: src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs:99]: On `SiloAddress`, we have a `bool Matches(SiloAddress other)` method to serve the purpose of "same identity but not identical" - maybe we should add that on `GrainAddress`

File: src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs (hunk 4 above)
Comment (benjaminpetit) [reply src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs:107]: Yes, but in another PR, I don't want to touch the logic from the in-cluster directory just now

**Review by ReubenBond — APPROVED** [review]

**Pull request merged**